File: heap/heap.cpp
#include<vector>

#ifndef HEAP
#define HEAP

template<class Key>
class Heap {
    private:
        // Props
        std::vector<Key> queue = {NULL};

        // Private methods
        void swap(int i, int j);

        void swim(int k);

        void sink(int k);

    public:
        // Constructor
        Heap();

        // Public methods
        int empty();

        int size();

        void insert(Key v);

        Key remove();

        Key max();
};

template<class Key>
void Heap<Key>::swap(int i, int j) {
    Key temp = queue[i];
    queue[i] = queue[j];
    queue[j] = temp;
}

template<class Key>
void Heap<Key>::swim(int k) {
    while (k > 1 && queue[k / 2] < queue[k]) {
        swap(k / 2, k);
        k = k / 2;
    }
}

template<class Key>
void Heap<Key>::sink(int k) {
    while (2 * k <= size()) {
        int j = 2 * k;
        if (j < size() && queue[j] < queue[j + 1]) {
            j++;
        }

        if (queue[k] >= queue[j]) {
            break;
        }

        swap(k, j);
        k = j;
    }
}

template<class Key>
Heap<Key>::Heap() = default;

template<class Key>
int Heap<Key>::empty() {
    return queue.size() <= 1;
}

template<class Key>
int Heap<Key>::size() {
    return queue.size() - 1;
}

template<class Key>
void Heap<Key>::insert(Key v) {
    queue.push_back(v);
    swim(size());
}

template<class Key>
Key Heap<Key>::remove() {
    Key max = queue[1];
    swap(1, queue.size() - 1);
    queue.pop_back();
    sink(1);

    return max;
}

template<class Key>
Key Heap<Key>::max() {
    return empty() ? NULL : queue[1];
}

#endif //HEAP

File: heap/heap.py
class Heap:
    def __init__(self) -> None:
        self.queue = [None]

    def __swap(self, i: int, j: int) -> None:
        Heap.__s_swap(self.queue, i, j)

    @staticmethod
    def __s_swap(arr: list, i: int, j: int) -> None:
        tmp = arr[i]
        arr[i] = arr[j]
        arr[j] = tmp

    def __swim(self, k: int) -> None:
        while k > 1 and self.queue[k // 2] < self.queue[k]:
            Heap.__s_swap(self.queue, k // 2, k)
            k = k // 2

    def __sink(self, k: int) -> None:
        Heap.__s_sink(self.queue, k, self.size())

    @staticmethod
    def __s_sink(arr: list, k: int, n: int):
        while 2 * k <= n:
            j = 2 * k
            if j < n and arr[j] < arr[j + 1]:
                j += 1

            if arr[k] >= arr[j]:
                break

            Heap.__s_swap(arr, k, j)
            k = j

    def empty(self) -> bool:
        return len(self.queue) <= 1

    def size(self) -> int:
        return len(self.queue) - 1

    def insert(self, v) -> None:
        self.queue.append(v)
        self.__swim(self.size())

    def remove(self):
        if self.empty():
            return None

        max_value = self.queue[1]
        self.__swap(1, self.size())
        self.queue.pop()
        self.__sink(1)

        return max_value

    def max(self):
        return None if self.empty() else self.queue[1]

    @staticmethod
    def sort(arr: list) -> None:
        n = len(arr)
        arr.insert(0, None)

        for k in range(n // 2, 0, -1):
            Heap.__s_sink(arr, k, n)

        while n > 1:
            Heap.__s_swap(arr, 1, n)
            n -= 1
            Heap.__s_sink(arr, 1, n)

        arr.pop(0)
